From f371e19203dc18315114aa94e57488de4f5a6bdc Mon Sep 17 00:00:00 2001
From: Pratham Patel <ppatel@ciq.com>
Date: Tue, 3 Jun 2025 12:11:19 +0530
Subject: [PATCH] netfilter: nf_tables: bail out on mismatching dynset and set
 expressions

jira VULN-679
cve CVE-2023-6622
commit-author Pablo Neira Ayuso <pablo@netfilter.org>
commit 3701cd390fd731ee7ae8b8006246c8db82c72bea

If dynset expressions provided by userspace is larger than the declared
set expressions, then bail out.

Fixes: 48b0ae046ee9 ("netfilter: nftables: netlink support for several set element expressions")
	Reported-by: Xingyuan Mo <hdthky0@gmail.com>
	Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>
(cherry picked from commit 3701cd390fd731ee7ae8b8006246c8db82c72bea)
	Signed-off-by: Pratham Patel <ppatel@ciq.com>
---
 net/netfilter/nft_dynset.c | 13 +++++++++----
 1 file changed, 9 insertions(+), 4 deletions(-)

diff --git a/net/netfilter/nft_dynset.c b/net/netfilter/nft_dynset.c
index f407b3e9c178..ae61505d4c41 100644
--- a/net/netfilter/nft_dynset.c
+++ b/net/netfilter/nft_dynset.c
@@ -274,10 +274,15 @@ static int nft_dynset_init(const struct nft_ctx *ctx,
 			priv->expr_array[i] = dynset_expr;
 			priv->num_exprs++;
 
-			if (set->num_exprs &&
-			    dynset_expr->ops != set->exprs[i]->ops) {
-				err = -EOPNOTSUPP;
-				goto err_expr_free;
+			if (set->num_exprs) {
+				if (i >= set->num_exprs) {
+					err = -EINVAL;
+					goto err_expr_free;
+				}
+				if (dynset_expr->ops != set->exprs[i]->ops) {
+					err = -EOPNOTSUPP;
+					goto err_expr_free;
+				}
 			}
 			i++;
 		}
-- 
2.47.1

